File: solutions/0019-remove-nth-node-from-end-of-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:

        # traverse once to calculate size of linked list

        size = 0
        head_ptr = head
        while head != None:
            head = head.next
            size += 1
        
        if size == 1:
            return None

        k = size - n # we need kth index from the head
        if k == 0:
            return head_ptr.next
            
        head_ptr2 = head_ptr
        for i in range(k):
            if i == k-1:
                head_ptr.next = head_ptr.next.next
            else:
                head_ptr = head_ptr.next

        return head_ptr2
        

File: solutions/0019-remove-nth-node-from-end-of-list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
        if (head == null) {
            return head;
        }
        int listSize = 0;
        ListNode headPtr = head;
        while (head != null) {
            head = head.next;
            listSize ++;
        }
        System.out.println("Size of linked list = " + listSize);

        // n >= 1 so if listSize == 1, we remove the only element available in the list
        if (listSize == 1) {
            return null;
        }

        // head to return, unless we remove the head itself - special case treated separately
        ListNode headPtr2 = headPtr;
        // reframing the indexToRemove from the start instead of from the end
        int indexToRemove = listSize - n;
        System.out.println("indexToRemove = " + indexToRemove);

        int i = 0;
        ListNode prev = null;
        while(headPtr != null) {
            // stop when you hit the index before indexToRemove
            if (i == indexToRemove) {
                if (prev == null) {
                    return headPtr.next;
                } else {
                    prev.next = headPtr.next;
                    break;
                }
            }
            prev = headPtr;
            headPtr = headPtr.next;
            i += 1;
        }
        return headPtr2;
    }
}
